### PR TITLE
[#19]: support rel=canonical url

### PR DESCRIPTION
**Pull request by vitkarpov**
Support for rel=canonical url.

Likely constructor already has an ability to specify another url in options, it's used if data-url is absent. I added a new method to utils `getCanonicalUrl`, it returns the url if link[rel=canonical] is in the document.

Also the code:

```js
utils.toArray(widgets)
    .forEach(likely);
```

doesn't work as expected 'cause likely function get 2s arguments: node and its index, and it's not the likely function expects — it should be node and options. Without adding canonical url the error couldn't be revealed.

### CODE REVIEW

**Comment by vitkarpov** [comment]
@iamakulov

File: source/index.js [hunk 1]
```diff
@@ -32,10 +37,12 @@ var likely = function (node, options) {
  * Initiate Likely buttons on load
  */
 likely.initiate = likely.initate = function () {
-    var widgets = dom.findAll('.' + config.name);
+    var nodes = dom.findAll('.' + config.name);
```
Comment (iamakulov) [inline source/index.js]: I’d better not do this. `widgets` is more specific and therefore more clear name.

File: source/index.js [hunk 2]
```diff
@@ -32,10 +37,12 @@ var likely = function (node, options) {
...
 
-    utils.toArray(widgets)
-        .forEach(likely);
+    utils.toArray(nodes)
+        .forEach(function(node) {
```
Comment (iamakulov) [inline source/index.js]: `function(` won’t pass ESLint.

File: source/utils.js [hunk 3]
```diff
@@ -199,6 +199,21 @@ var utils = {
 
         object[last] = value;
     },
+
+    /**
+     * Returns href from <link rel='canonical'>
+     * if presents in the document
+     *
+     * @returns {String?}
```
Comment (iamakulov) [inline source/utils.js]: Is this notation valid? I’ve thought `String|null` should be used in such cases.

File: source/index.js [hunk 4]
```diff
@@ -19,6 +19,11 @@ var likely = function (node, options) {
         widget.update(fullOptions);
     }
     else {
+        var canonicalUrl = utils.getCanonicalUrl();
+
+        if (canonicalUrl) {
+            fullOptions.url = canonicalUrl;
+        }
```
Comment (iamakulov) [inline source/index.js]: Doesn’t this override the url passed from the options? We want the url in the options be more specific than the canonical url, and the canonical url be more specific than the page url.

Also, I’d better move this to https://github.com/ilyabirman/Likely/blob/master/source/index.js#L50, where the page URL is extracted. There can only be one canonical URL per page, so it belongs rather to Likely defaults than to its constructor.

**Comment by iamakulov** [comment]
@vitkarpov, added a few comments.

> Also the code... doesn't work as expected

Good catch, thank you!

Also, two points:

* Could you please revert the changes in the `release/likely.*` files? We don’t want to update the release files in PRs because it could cause merge conflicts where there’re several simultaneous PRs.
* Could you please split your changes into two commits: one with the `utils.toArray(widgets).forEach(likely);` fix, and another with the canonical URL implementation – and force-push them? This way we could keep our commit history cleaner as each commit will only do one specific change.

**Comment by iamakulov** [comment]
@vitkarpov, and a third point: until we have a CI, could you please run `npm run codestyle` on your own before committing?

File: source/utils.js (hunk 3 above)
Comment (vitkarpov) [inline source/utils.js]: There's a [nullable type](https://developers.google.com/closure/compiler/docs/js-for-compiler?hl=de#types), but I misused it — the question mark should be before the type, not after.

**Comment by vitkarpov** [comment]
@iamakulov Done! Considered all the remarks.

**Comment by iamakulov** [comment]
@vitkarpov, thank you, this looks great!

You missed one point:

> Could you please split your changes into two commits: one with the `utils.toArray(widgets).forEach(likely);` fix, and another with the canonical URL implementation – and force-push them? This way we could keep our commit history cleaner as each commit will only do one specific change.

Once it’s done, I’m ready to check the PR locally and merge it. Or feel free to comment if you doubt in this point.

**Comment by vitkarpov** [comment]
> You missed one point

Oh, right.

**Comment by vitkarpov** [comment]
Done

**Comment by iamakulov** [comment]
Works great, thank you!

**Pull request merged**